rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseInsertExecutor.java

```diff
@@ -0,0 +1,300 @@
+                break;
+            }
+        }
+        return pkIndex;
+    }
+
+    /**
+     * parse primary key value from statement.
+     * @return
+     */
+    protected List<Object> parsePkValuesFromStatement() {
+        // insert values including PK
+        SQLInsertRecognizer recognizer = (SQLInsertRecognizer) sqlRecognizer;
+        final int pkIndex = getPkIndex();
+        if (pkIndex == -1) {
+            throw new ShouldNeverHappenException(String.format("pkIndex is %d", pkIndex));
+        }
+        List<Object> pkValues = null;
+        if (statementProxy instanceof PreparedStatementProxy) {
+            PreparedStatementProxy preparedStatementProxy = (PreparedStatementProxy) statementProxy;
+
+            List<List<Object>> insertRows = recognizer.getInsertRows();
+            if (insertRows != null && !insertRows.isEmpty()) {
+                ArrayList<Object>[] parameters = preparedStatementProxy.getParameters();
+                final int rowSize = insertRows.size();
+
+                if (rowSize == 1) {
+                    Object pkValue = insertRows.get(0).get(pkIndex);
+                    if (PLACEHOLDER.equals(pkValue)) {
+                        pkValues = parameters[pkIndex];
+                    } else {
+                        pkValues = insertRows.stream().map(insertRow -> insertRow.get(pkIndex)).collect(Collectors.toList());
+                    }
+                } else {
+                    int totalPlaceholderNum = -1;
+                    pkValues = new ArrayList<>(rowSize);
+                    for (int i = 0; i < rowSize; i++) {
+                        List<Object> row = insertRows.get(i);
+                        // oracle insert sql statement specify RETURN_GENERATED_KEYS will append :rowid on sql end
+                        // insert parameter count will than the actual +1
+                        if (row.isEmpty()) {
+                            continue;
+                        }
+                        Object pkValue = row.get(pkIndex);
+                        int currentRowPlaceholderNum = -1;
+                        for (Object r : row) {
+                            if (PLACEHOLDER.equals(r)) {
+                                totalPlaceholderNum += 1;
+                                currentRowPlaceholderNum += 1;
+                            }
+                        }
+                        if (PLACEHOLDER.equals(pkValue)) {
+                            int idx = pkIndex;
+                            if (i != 0) {
+                                idx = totalPlaceholderNum - currentRowPlaceholderNum + pkIndex;
+                            }
+                            ArrayList<Object> parameter = parameters[idx];
+                            pkValues.addAll(parameter);
+                        } else {
+                            pkValues.add(pkValue);
+                        }
+                    }
+                }
+            }
```

Comment:
Reduce coupling and implement in 3 steps
1.getInsertRows
2.get pkIndex list from rows
3.get pkValue list from parameters